### PR TITLE
create {sbox,ctr}factory

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

> /kind api-change
> /kind bug
> /kind ci
> /kind cleanup
> /kind dependency-change
> /kind deprecation
/kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake
> /kind other

#### What this PR does / why we need it:
the existence of two packages with duplicated names has been a source of confusion for new developers.
The original purpose of the pkg/{contanier,sandbox} packages were supposed to replace the oci/container and lib/sandbox packages. but that never really panned out, and the functionality moved more to a container/sandbox factory.

Now, we make explicit the design of server->*factory->internal structure.

Also, drop the interfaces in those packages. We don't actually use them as interfaces, and it is not a free abstraction.

#### Special notes for your reviewer:
fixes https://github.com/cri-o/cri-o/issues/3965
#### Does this PR introduce a user-facing change?

```release-note
none
```

### CODE REVIEW

**Comment by haircommander** [comment]
carrying https://github.com/cri-o/cri-o/pull/4623 and https://github.com/cri-o/cri-o/pull/4623. I can do this first if it's preferred

**Comment by haircommander** [comment]
I think this deserves explicit ack from most maintainers here
@kolyshkin @mrunalp @saschagrunert @umohnani8 @QiWang19 @wgahnagl  PTAL

**Review by saschagrunert — APPROVED** [review]
LGTM

File: pkg/sandbox/infra.go [hunk 1]
```diff
@@ -68,3 +85,100 @@ func PauseCommand(cfg *libconfig.Config, image *v1.Image) ([]string, error) {
...
+	}
+
+	dnsServers := s.config.DNSConfig.Servers
+	dnsSearches := s.config.DNSConfig.Searches
+	dnsOptions := s.config.DNSConfig.Options
+	s.resolvPath = fmt.Sprintf("%s/resolv.conf", podContainer.RunDir)
+	err := ParseDNSOptions(dnsServers, dnsSearches, dnsOptions, s.resolvPath)
+	defer func() {
+		if retErr != nil {
+			if err := os.RemoveAll(s.resolvPath); err != nil {
```
Comment (kolyshkin) [inline pkg/sandbox/infra.go]: This should be `os.Remove` (as no recursion is intended) + ignoring `os.IsNotExist` error (as os.Remove can return it).

File: pkg/sandbox/infra.go [hunk 2]
```diff
@@ -37,6 +50,10 @@ func (s *sandbox) InitInfraContainer(serverConfig *libconfig.Config, podContaine
 	}
 	g.SetProcessArgs(pauseCommand)
 
+	if err := s.createResolvPath(podContainer); err != nil {
```
Comment (kolyshkin) [inline pkg/sandbox/infra.go]: nit: I would call it `createResolvConf` (not `createResolvPath`).

File: pkg/sandbox/infra.go [hunk 3]
```diff
@@ -68,3 +85,100 @@ func PauseCommand(cfg *libconfig.Config, image *v1.Image) ([]string, error) {
...
+func ParseDNSOptions(servers, searches, options []string, path string) (retErr error) {
+	nServers := len(servers)
+	nSearches := len(searches)
+	nOptions := len(options)
+	if nServers == 0 && nSearches == 0 && nOptions == 0 {
+		return copyFile("/etc/resolv.conf", path)
+	}
+
+	if nSearches > maxDNSSearches {
+		return fmt.Errorf("DNSOption.Searches has more than %d domains", maxDNSSearches)
```
Comment (kolyshkin) [inline pkg/sandbox/infra.go]: :+1: for fixing the hardcoded `6`.

File: pkg/sandbox/infra.go [hunk 4]
```diff
@@ -68,3 +85,100 @@ func PauseCommand(cfg *libconfig.Config, image *v1.Image) ([]string, error) {
...
+
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if nSearches > 0 {
+		data := fmt.Sprintf("search %s\n", strings.Join(searches, " "))
+		_, err = f.Write([]byte(data))
```
Comment (kolyshkin) [inline pkg/sandbox/infra.go]: I'd do
```go
_, err = f.WriteString("search " + strings.Join(searches, " ") + "\n")
```

File: pkg/sandbox/infra.go [hunk 5]
```diff
@@ -68,3 +85,100 @@ func PauseCommand(cfg *libconfig.Config, image *v1.Image) ([]string, error) {
...
+		if err != nil {
+			return err
+		}
+	}
+
+	if nServers > 0 {
+		data := fmt.Sprintf("nameserver %s\n", strings.Join(servers, "\nnameserver "))
+		_, err = f.Write([]byte(data))
```
Comment (kolyshkin) [inline pkg/sandbox/infra.go]: ditto

File: pkg/sandbox/infra.go [hunk 6]
```diff
@@ -68,3 +85,100 @@ func PauseCommand(cfg *libconfig.Config, image *v1.Image) ([]string, error) {
...
+		if err != nil {
+			return err
+		}
+	}
+
+	if nOptions > 0 {
+		data := fmt.Sprintf("options %s\n", strings.Join(options, " "))
+		_, err = f.Write([]byte(data))
```
Comment (kolyshkin) [inline pkg/sandbox/infra.go]: ditto

File: internal/ctrfactory/ctrfactory.go [hunk 7]
```diff
@@ -1,4 +1,4 @@
-package container
+package ctrfactory
```
Comment (kolyshkin) [inline internal/ctrfactory/ctrfactory.go:1]: (GH can't let me comment on the commit so I'm commenting here)

The commit message should be self-explanatory. `ctrfactory: do the same as sboxfactory` is not.

I'd say something like
```
pkg/container: rename to ctrfactory

For reasons behind this, see earlier commit titled "XXXX"
```

or just copy-paste the reasoning as it's not too long.

File: internal/ctrfactory/ctrfactory.go [hunk 8]
```diff
@@ -553,9 +558,35 @@ func (c *container) SpecSetProcessArgs(imageOCIConfig *v1.Image) error {
 	return nil
 }
 
+// SpecAddNamespaces adds the specified namespace paths
```
Comment (kolyshkin) [inline internal/ctrfactory/ctrfactory.go:481]: Looks like there's some mishap -- this should not belong to this (renaming) commit.

**Comment by kolyshkin** [comment]
Kudos for dropping interfaces with a single implementation!

Maybe if you drop interfaces first and rename second, there will be less changes to look into -- but I'm fine as it is.

Left a few minor nits.

**Comment by haircommander** [comment]
> Kudos for dropping interfaces with a single implementation!
> 
> Maybe if you drop interfaces first and rename second, there will be less changes to look into -- but I'm fine as it is.
> 
> Left a few minor nits.

that definitely would have been less, but I am not sufficiently motivated to undo it now :laughing: thanks for the notes, addressed.

**Comment by haircommander** [comment]
closing in favor of https://github.com/cri-o/cri-o/pull/5623

**Pull request closed without merging**